compiler/rustc_resolve/src/macros.rs

```diff
@@ -140,10 +142,13 @@ pub(crate) fn registered_tools(tcx: TyCtxt<'_>, (): ()) -> RegisteredTools {
             }
         }
     }
-    // We implicitly add `rustfmt` and `clippy` to known tools,
+    // We implicitly add `rustfmt`, `clippy`, `diagnostic` to known tools,
     // but it's not an error to register them explicitly.
     let predefined_tools = [sym::clippy, sym::rustfmt];
     registered_tools.extend(predefined_tools.iter().cloned().map(Ident::with_dummy_span));
+    if tcx.features().diagnostic_namespace {
```

Comment:
Feature gates should not change any behavior, only produce errors.
`diagnostic` should always be added to registered tools, and we need to produce the feature error when it's actually used.